plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java

```diff
@@ -109,29 +98,33 @@ private boolean canCoerceForMap(HiveType fromHiveType, HiveType toHiveType)
         HiveType fromValueType = HiveType.valueOf(((MapTypeInfo) fromHiveType.getTypeInfo()).getMapValueTypeInfo().getTypeName());
         HiveType toKeyType = HiveType.valueOf(((MapTypeInfo) toHiveType.getTypeInfo()).getMapKeyTypeInfo().getTypeName());
         HiveType toValueType = HiveType.valueOf(((MapTypeInfo) toHiveType.getTypeInfo()).getMapValueTypeInfo().getTypeName());
-        return (fromKeyType.equals(toKeyType) || canCoerce(fromKeyType, toKeyType)) &&
-                (fromValueType.equals(toValueType) || canCoerce(fromValueType, toValueType));
+        return (fromKeyType.equals(toKeyType) || canCoerce(fromKeyType, toKeyType, hiveTimestampPrecision)) &&
+                (fromValueType.equals(toValueType) || canCoerce(fromValueType, toValueType, hiveTimestampPrecision));
     }
 
-    private boolean canCoerceForList(HiveType fromHiveType, HiveType toHiveType)
+    private boolean canCoerceForList(HiveType fromHiveType, HiveType toHiveType, HiveTimestampPrecision hiveTimestampPrecision)
     {
         if (fromHiveType.getCategory() != Category.LIST || toHiveType.getCategory() != Category.LIST) {
             return false;
         }
         HiveType fromElementType = HiveType.valueOf(((ListTypeInfo) fromHiveType.getTypeInfo()).getListElementTypeInfo().getTypeName());
         HiveType toElementType = HiveType.valueOf(((ListTypeInfo) toHiveType.getTypeInfo()).getListElementTypeInfo().getTypeName());
-        return fromElementType.equals(toElementType) || canCoerce(fromElementType, toElementType);
+        return fromElementType.equals(toElementType) || canCoerce(fromElementType, toElementType, hiveTimestampPrecision);
     }
 
-    private boolean canCoerceForStruct(HiveType fromHiveType, HiveType toHiveType)
+    private boolean canCoerceForStructOrUnion(HiveType fromHiveType, HiveType toHiveType, HiveTimestampPrecision hiveTimestampPrecision)
     {
-        if (fromHiveType.getCategory() != Category.STRUCT || toHiveType.getCategory() != Category.STRUCT) {
+        if ((fromHiveType.getCategory() != Category.STRUCT && fromHiveType.getCategory() != Category.UNION)
+                || (toHiveType.getCategory() != Category.STRUCT && toHiveType.getCategory() != Category.UNION)) {
             return false;
         }
-        List<String> fromFieldNames = ((StructTypeInfo) fromHiveType.getTypeInfo()).getAllStructFieldNames();
-        List<String> toFieldNames = ((StructTypeInfo) toHiveType.getTypeInfo()).getAllStructFieldNames();
-        List<HiveType> fromFieldTypes = extractStructFieldTypes(fromHiveType);
-        List<HiveType> toFieldTypes = extractStructFieldTypes(toHiveType);
+        HiveType fromHiveTypeStruct = (fromHiveType.getCategory() == ObjectInspector.Category.UNION) ? HiveType.toHiveType(fromHiveType.getType(typeManager, hiveTimestampPrecision)) : fromHiveType;
```

Comment:
nit: since this is invoked multiple times - may be have separate getStructFromUnion

Reply:
Suggest name as convertUnionToStruct

Reply:
getStructFromUnion is a name seems like union is a collections of properties and getting one of the struct property from the union.